### DISCUSSION TITLE
How should I be handling storing precreated shapes instead of building them at runtime?

### DISCUSSION THREAD

**Discussion by Pindrought** [jrouwe/JoltPhysics#728] [Q&A] [answered]
I have my own custom 3d format, and currently I have it set up where I am building the binary state for the mesh shape for my static meshes.

Here is a basic example of what i'm doing at the moment in the pre processing when I build my custom file data...
```cpp
JPH::RegisterDefaultAllocator();
JPH::VertexList verts;
JPH::IndexedTriangleList triangles;

//INSERT CODE HERE TO BUILD THE VertexList / IndexedTriangleList

JPH::MeshShapeSettings settings(verts, triangles);
auto shapeResult = settings.Create();
if (!shapeResult.IsValid())
{
	return false;
}

StreamOutToMem mem; //StreamOut implementation so that I can get the bytes/size of data in an array to save it
shapeResult.Get()->SaveBinaryState(mem);
auto data = mem.GetData();
auto size = mem.GetDataSize();
//after I have the byte chunk data/size, I store it in my custom file format and retrieve it when parsing a model
```

Here is the StreamOutToMem impl
```cpp
class StreamOutToMem : public JPH::StreamOut
{
public:
	void Clear() { data.clear(); }
	const byte* GetData() { return data.data(); }
	const int GetDataSize() { return data.size(); }
	/// Write a string of bytes to the binary stream
	virtual void WriteBytes(const void* inData, size_t inNumBytes) override 
	{ 
		int offset = data.size();
		data.resize(data.size() + inNumBytes);
		memcpy(&data[offset], inData, inNumBytes);
	}

	/// Returns true if there was an IO failure
	virtual bool IsFailed() const override { return false; } //I don't think this makes sense with the way i'm using it, so i'm just returning false?

private:
	std::vector<byte> data;
};
```

Honestly it feels like the whole approach i'm doing here is wrong, but I didn't want to be building these shapes at runtime if I can do it during pre processing. 

When I have the shape binary data, how should I be restoring this to a shape? I was planning on putting in an impl to StreamIn and calling RestoreBinaryState(), but then I saw [the docs say to not call this function.](https://jrouwe.github.io/JoltPhysics/class_shape.html#a98a42fc4c28896d87ac4287c38f213ad)

I see the docs say to call **JPH::Shape::sRestoreWithChildren()**.

I tried doing this when loading in a model...
```cpp
JPH::Shape::IDToShapeMap id_to_shape;
JPH::Shape::IDToMaterialMap id_to_material;
StreamInWrapper stream_in(parser.m_ColliderGroups[0].Colliders[0].Bytes.data(), 
							parser.m_ColliderGroups[0].Colliders[0].Bytes.size());
JPH::Shape::ShapeResult result = JPH::Shape::sRestoreWithChildren(stream_in, id_to_shape, id_to_material);

if (result.IsValid())
	pModel->m_StaticShape = result.Get();
```
With the following StreamInWrapper impl
```
class StreamInWrapper : public JPH::StreamIn
{
public:
	/// Constructor
	StreamInWrapper(void* data, size_t chunkSize)
	{
		this->data = data;
		this->chunkSize = chunkSize;
	}

	/// Write a string of bytes to the binary stream
	virtual void		ReadBytes(void* outData, size_t inNumBytes) override 
	{
		if ((offset + inNumBytes) > chunkSize)
		{
			memcpy(outData, (byte*)(data) + offset, chunkSize - offset);
			offset = chunkSize;
		}
		else
		{
			memcpy(outData, (byte*)(data) + offset, inNumBytes);
			offset += inNumBytes;
		}
	}

	/// Returns true when an attempt has been made to read past the end of the file
	virtual bool		IsEOF() const override 
	{ 
		if (offset > chunkSize)
		{
			return true;
		}
		return false;
	}

	/// Returns true if there was an IO failure
	virtual bool		IsFailed() const override 
	{ 
		if (offset > chunkSize)
		{
			return true;
		}
		return false;
	}

private:
	void* data = nullptr;
	size_t chunkSize = 0;
	size_t offset = 0;
};
```
And I am running into this issue...
![image](https://github.com/jrouwe/JoltPhysics/assets/39708313/fdaf10c5-2b68-4d3b-9f8f-d18c8ba4342e)

Edit: I notice that if I use **sRestoreFromBinaryState()** instead of **sRestoreWithChildren()** that it works as expected. I am now wondering why I get an error when using **sRestoreWithChildren()**.

**Comment by jrouwe** [comment, marked as answer]
> I notice that if I use sRestoreFromBinaryState() instead of sRestoreWithChildren() that it works as expected. I am now wondering why I get an error when using sRestoreWithChildren().

Because you need to pair SaveWithChildren with sRestoreWithChildren and SaveBinaryState with sRestoreFromBinaryState. The first pair of functions is there to save everything, the second pair only saves the shape and is a lower level API to be used to share shapes between different streams.

There is a section on serialization in [section Saving Shapes](https://jrouwe.github.io/JoltPhysics/) and there's more documentation in the code [here](https://github.com/jrouwe/JoltPhysics/blob/e54a90192265e46193d004d8942f8bd0156c6c74/Jolt/Physics/Collision/Shape/Shape.h#L352-L356).

**Reply by Pindrought** [reply]
That makes perfect sense. Thanks again so much jrouwe!!
